File: Python/Sorting/radix_sort.py
import math

def radix_sort(arr):
	# print len(arr)
	max_num = max(arr)
	# print max_num
	num_of_digits = len(str(max_num))
	# print num_of_digits
	
	for i in range(0,num_of_digits):
		index_count = []
		result = []

		for j in range(0,10):
			index_count.append(0)

		for l in range(0,len(arr)):
			result.append(0)
			index_count[(arr[l]/(int(math.pow(10,i))))%10] += 1
		# print index_count

		for k in range(1,10):
			index_count[k] += index_count[k-1]
		# print index_count

		for m in range(len(arr)-1,-1,-1):
			result[index_count[(arr[m]/(int(math.pow(10,i))))%10] - 1] = arr[m]
			index_count[(arr[m]/(int(math.pow(10,i))))%10] -= 1
		
		arr = result

	return result

if __name__ == '__main__':
	arr = [7,4,9,6,5,3,19,37,1,7,4,6,5]
	print 'Original: ',arr
	sorted_arr = radix_sort(arr)
	print 'Sorted: ',sorted_arr

File: Java/Sorting/RadixSort.java
package soritng;
import java.util.Arrays;

public class RadixSort {
	public static void main(String[] args) {
		int[] arr = {7,4,9,6,5,3,19,37,1,7,4,6,40};
		System.out.println("Original Array: "+Arrays.toString(arr));

		int[] sorted_arr = radix_sort(arr);
		System.out.println("Sorted Array: "+Arrays.toString(sorted_arr));
	}

	public static int[] radix_sort(int[] arr) {
		int max_num = getMax(arr);
		// System.out.println(""+max_num);
		int num_of_digits = (Integer.toString(max_num)).length();
		// System.out.println(""+num_of_digits);

		for ( int i=0;i<num_of_digits;i++ ) {
			int[] index_count = new int[10];
			int[] result = new int[arr.length];

			for ( int j=0;j<10;j++ ) {
				index_count[j] = 0;
			}

			for ( int k=0;k<arr.length;k++ ) {
				result[k] = 0;
				index_count[(arr[k]/((int)(Math.pow(10,i))))%10] += 1;
			}
			// System.out.println(Arrays.toString(index_count));

			for ( int l=1;l<10;l++ ) {
				index_count[l] += index_count[l-1];
			}
			// System.out.println(Arrays.toString(index_count));

			for ( int m=arr.length-1;m>=0;m-- ) {
				result[index_count[(arr[m]/((int)(Math.pow(10,i))))%10] - 1] = arr[m];
				index_count[(arr[m]/((int)(Math.pow(10,i))))%10] -= 1;
			}

			arr = result;
		}
		return arr;
	}

	public static int getMax(int[] arr) {
		int max_num = arr[0];
		for ( int i=0;i<arr.length;i++ ) {
			if ( arr[i] > max_num ) {
				max_num = arr[i];
			}
		}

		return max_num;
	}
}